File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*; 
class Solution {
    public boolean solution(String[] phone_book) {
        // 문자열의 길이로 오름차순 정렬 
        Arrays.sort(phone_book, (o1, o2) -> o1.length() - o2.length());
        
        Map<String, Integer> map = new HashMap<>(); 
        for(String str: phone_book){
            // System.out.println("about " + str);
            // System.out.println(map.keySet());
            
            for(int i=1; i<=str.length(); i++){
                // System.out.println(str.substring(0, i));
                if(map.containsKey(str.substring(0, i))) {
                    return false;  
                } 
            }
            map.putIfAbsent(str, 0);
        }
        return true;
    }
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>
#include <vector>
#include <iostream> 
#include <algorithm> 
using namespace std;

bool solution(vector<string> phone_book) {
    bool answer = true;
    
    sort(phone_book.begin(), phone_book.end()); 
    
    for(int i=0; i<phone_book.size()-1; i++){
        // for(int j=i+1; j<phone_book.size(); j++){ 
        // 앞의 문자열이 뒤의 문자열의 접두사인지만 확인하면 된다. 
        if(phone_book[i+1].find(phone_book[i]) == 0){ 
            return false; 
        }
        // }
    }
    return true;
}
/*
접두어 

사전순 정렬
123 < 1234 

정렬을 먼저 하고 
앞에 있는 게 뒤에 있는 것들 중에서 찾았을 때 그 위치가 0이면 접두어겠네 

*/